### PR TITLE
undo redo history

### PR DESCRIPTION
**Pull request by Yuddomack**
hi! I tried working on it.
but I did this task without fully understand, so need to some your feedback

thank you!

### CODE REVIEW

File: packages/core/lib/use-puck-data.ts
Comment (chrisvxd) [inline packages/core/lib/use-puck-data.ts:136]: Hm I'm a bit unsure about this particular file. Perhaps we would be be better off going with deep object diff library, and store the delta, rather than re-calling the dispatchers here.

You could use [deep-diff](https://www.npmjs.com/package/deep-diff) and the [applyChange API](https://github.com/flitbit/diff/tree/e2b5b86add402ae4373ab025ee7e9a98e3424023#changes), or try out [rfc6902](https://github.com/chbrown/rfc6902). I don't think it really matters which.

Example using `deepDiff` inside `reducer.ts`, without modifying existing reducer code to avoid polluting the state machine

```tsx

const storeHistory = (reducer: StateReducer) => {
  const history: any = [];

  return (data, action) => {
    const newData = reducer(data, action);

    const diff = deepDiff.diff(data, newData);

    history.push(diff);

    return newData;
  };
};

export const createReducer = ({ config }: { config: Config }): StateReducer =>
  storeHistory((data, action) => {
    if (action.type === "insert") {
      const emptyComponentData = {
        type: action.componentType,
        props: {
          ...(config.components[action.componentType].defaultProps || {}),
          id: generateId(action.componentType),
        },
      };
   ...
```

**Review by chrisvxd — CHANGES_REQUESTED** [review]
Thanks for the big effort again @Yuddomack! 

I've left a suggestion for how to simplify the history recording using a diffing library.

File: packages/core/lib/reducer.ts
Comment (Yuddomack) [reply packages/core/lib/reducer.ts:51]:
```javascript
// before data
data = [
  { type: "Hero", ... },
  { type: "VerticalSpace", ... },
  { type: "Heading", ... },
  ...
]

// data after removed 'Hero' // dispatch({ type: "remove", index, zone })
data = [
  { type: "VerticalSpace", ... },
  { type: "Heading", ... },
  ...
]
```

umm I thought What information do I need to restore what was removed?
-> zone, index, (before removed) data

the data could be extracted using `getItem` when excuted remove action.
but I needed the ability to put it back in its original location. (`insert action` is based on new data object(emptyComponentData), it is judged that it cannot be reused)

so I added an implementation for the recovery action.
(Perhaps this idea should be discussed on https://github.com/measuredco/puck/pull/111#discussion_r1337537243 first.)

File: packages/core/lib/use-puck-data.ts
Comment (Yuddomack) [reply packages/core/lib/use-puck-data.ts:136]: Oh this idea looks really cool!
I'll come back using this ⛹️‍♂️

Comment (Yuddomack) [reply packages/core/lib/use-puck-data.ts]: Hi! @chrisvxd
I tried applying deep-diff. I still have a little considered about each responsibility, but it ready to review 🔥

File: packages/core/lib/__tests__/use-action-history.spec.tsx
Comment (chrisvxd) [inline packages/core/lib/__tests__/use-action-history.spec.tsx:115]:
```suggestion
    test("canRewind should be true until the start of the history is reached", () => {
```

File: packages/core/lib/__tests__/use-action-history.spec.tsx
Comment (chrisvxd) [inline packages/core/lib/__tests__/use-action-history.spec.tsx:125]:
```suggestion
  describe("when multiple action histories are recorded, rewound and then fast-forwarded", () => {
```

File: packages/core/lib/__tests__/use-action-history.spec.tsx
Comment (chrisvxd) [inline packages/core/lib/__tests__/use-action-history.spec.tsx:156]:
```suggestion
    test("canForward should be true until the end of the history is reached", () => {
```

File: packages/core/lib/__tests__/use-action-history.spec.tsx
Comment (chrisvxd) [inline packages/core/lib/__tests__/use-action-history.spec.tsx:185]:
```suggestion
    test("the forward history is overridden when recording a new action ", () => {
```

File: packages/core/lib/use-puck-history.ts
Comment (chrisvxd) [inline packages/core/lib/use-puck-history.ts:30]: This is excellent. What do you think about adding tests here?

No worries if not - we don't have them anywhere else yet.

File: packages/core/lib/__tests__/use-action-history.spec.tsx
Comment (chrisvxd) [inline packages/core/lib/__tests__/use-action-history.spec.tsx:192]: Could you add this to the CI file so the tests run as part of this build?

```yml
      - name: Run tests
        run: yarn test
```

`.github/workflows/ci.yml` around L30 would be great

**Review by chrisvxd — CHANGES_REQUESTED** [review]
This is really excellent work! Thank you 🙏

Very happy with this approach. Most of my feedback is on phrasing in the unit tests.

File: packages/core/lib/use-puck-history.ts
Comment (Yuddomack) [reply packages/core/lib/use-puck-history.ts:55]: If the user quickly modifies data, each action is recorded.

like this

```
quickly typing) This page was built with Puck -> This page was built with Puck1234
rewind) This page was built with Puck123
rewind) This page was built with Puck12
rewind) This page was built with Puck1
rewind) This page was built with Puck
```

and I thought that recording of actions at once was good for user experience. (also memory of browser)

```
quickly typing) This page was built with Puck -> This page was built with Puck1234
rewind) This page was built with Puck
```

that's it!

File: packages/core/lib/use-puck-history.ts
Comment (Yuddomack) [reply packages/core/lib/use-puck-history.ts:30]: When #89 or #90 is completed, I will try with refer to that!

I think this area would be meaningful if it were basically at the level of mocking the reducer or integration testing of the puck component, but I need a better understanding of your testing philosophy.

If so, I can also help write tests for other components (or even e2e with demo app).

**Comment by Yuddomack** [comment]
Hi! @chrisvxd Your request has arrived!
- apply shortcut with each platform ([The meta key is the same as cmd on macOS and os key on Windows in react-hotkeys-hook](https://react-hotkeys-hook.vercel.app/docs/documentation/useHotkeys/basic-usage))
- resolve conflicts
- test descriptions + ci.yml
- some comments (for test, debounce)

File: packages/core/lib/use-puck-history.ts
Comment (chrisvxd) [reply packages/core/lib/use-puck-history.ts:55]: Right, perfect. Thanks for clarifying!

File: packages/core/package.json
Comment (chrisvxd) [inline packages/core/package.json:11]: Oops, we need to add this to the root `package.json` (`turbo run test`) for the CI script to work! Would you add mind adding that too, please?

File: packages/core/lib/use-puck-history.ts
Comment (chrisvxd) [reply packages/core/lib/use-puck-history.ts:30]: That makes sense!

#89 will be unit tests, #90 more integration but I'm unclear on the strategy yet.

I'm working on #89 next. We'll pick this up afterwards.

File: packages/core/package.json
Comment (chrisvxd) [inline packages/core/package.json:41]: This looks like a merge conflict - react-feather should have been removed in favour of the `devDependency` (see #121)

File: packages/core/lib/use-puck-history.ts
Comment (chrisvxd) [inline packages/core/lib/use-puck-history.ts:26]: Sorry, on macOS redo is actually CMD + SHIFT + Z.

Perhaps we should split this into two hotkeys?

```suggestion
  useHotkeys("cmd+shift+z", forward, { preventDefault: true });
  useHotkeys("ctrl+y", forward, { preventDefault: true });
```

or just define meta for both - I don't think it hurts

```suggestion
  useHotkeys("meta+shift+z", forward, { preventDefault: true });
  useHotkeys("meta+y", forward, { preventDefault: true });
```

(untested examples)

**Review by chrisvxd — CHANGES_REQUESTED** [review]
Wonderful! A couple more tiny comments before approval.

We're planning on doing v0.8.0@latest release soon, so will hold off merging this until _after_ that goes in to give it a chance to be tested in canary.

File: packages/core/package.json
Comment (chrisvxd) [reply packages/core/package.json:11]: You'll also need to add this to `turbo.json`

```
"test": {},
```

File: packages/core/lib/use-puck-history.ts
Comment (chrisvxd) [reply packages/core/lib/use-puck-history.ts:55]: Could we actually increase the DEBOUNCE_TIME here?

Maybe 250ms would be okay. `100` feels like it's still recording a lot of what I'm typing.

File: packages/core/lib/use-puck-history.ts
Comment (Yuddomack) [reply packages/core/lib/use-puck-history.ts:26]: cool! I like second one! (I checked it work)

**Comment by Yuddomack** [comment]
Hi! @chrisvxd 
Additional requests are also prepared! (maybe? 😆)

File: packages/core/lib/use-puck-history.ts
Comment (chrisvxd) [reply packages/core/lib/use-puck-history.ts:56]: Much better!

**Review by chrisvxd — APPROVED** [review]
🚀 Epic work @Yuddomack!

**Comment by chrisvxd** [comment]
@Yuddomack available in `0.9.0-canary.222697e`!

**Comment by chrisvxd** [comment]
@Yuddomack would love to have you in the Discord, if you're interested: https://discord.com/invite/D9e4E3MQVZ

**Comment by Yuddomack** [comment]
@chrisvxd cool! I joined it

**Pull request merged**
